### PR TITLE
subctl diagnose: Add validation for service-discovery

### LINKED ISSUE #50: subctl diagnose: Add validation for service-discovery

### PR DESCRIPTION
**Pull request by LiavSlama**
This commit adds support for Lighthouse diagnose

Fixes: https://github.com/submariner-io/subctl/issues/50
‏
Signed-off-by: LiavSlama <64088815+LiavSlama@users.noreply.github.com>
and itay62 <itay62@gmail.com>

### CODE REVIEW

**Comment by yboaron** [comment]
cc @vthapar

**Comment by yboaron** [comment]
@LiavSlama build fails, you should set **servicediscovery**  file as GO file

File: cmd/subctl/diagnose.go
Comment (yboaron) [inline cmd/subctl/diagnose.go]: Lighthouse

File: cmd/subctl/diagnose.go
Comment (tpantelis) [inline cmd/subctl/diagnose.go]:
```suggestion
		Use:   "service-discovery",
		Short: "Check service discovery functionality",
		Long:  "This command checks if service discovery is functioning properly.",
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
// This function checks if all ServiceExports have a matching ServiceImport and if an EndpointSlice has been created for the service.
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
	status.Start("Checking if services have been exported properly")
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
	status.Success("All services have been exported properly")
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go:117]: Checking for the existence of the `ServiceImport` is ok but we really want to know why it wasn't created. Also a service isn't completely exported until it's been synced to the broker. So I think it would really be useful to check that the `mcsv1a1.ServiceExportValid` condition type exists in the `ServiceExport` status and the condition's `Status` field is true. If not, then report the message. @vthapar @aswinsuryan WDYT

File: cmd/subctl/diagnose.go
Comment (tpantelis) [inline cmd/subctl/diagnose.go]: We really want to check if service discovery is installed.

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
			if apierrors.IsNotFound(err) {
				status.Failure("No EndpointSlice found for exported service %s/%s", ns, name)
			} else {
				status.Failure("Error retrieving EndPointSlice for exported service %s/%s", ns, name)
				return
			}
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]: This really isn't necessary - an overall success message is already logged after this.

```suggestion
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
			if apierrors.IsNotFound(err) {
				status.Failure("No ServiceImport found for exported service %s/%s", ns, name)
			} else {
				status.Failure("Error retrieving ServiceImport for exported service %s/%s", ns, name)
			}

			foundAll = false
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: cmd/subctl/diagnose.go
Comment (LiavSlama) [reply cmd/subctl/diagnose.go:165]: Thanks @tpantelis! So this instead of what you wrote above?

Comment (tpantelis) [reply cmd/subctl/diagnose.go]: I pushed https://github.com/submariner-io/subctl/pull/234.

Comment (LiavSlama) [reply cmd/subctl/diagnose.go:165]: You mean I should use it for this comment @tpantelis?

Comment (LiavSlama) [reply cmd/subctl/diagnose.go]: Thanks

Comment (tpantelis) [reply cmd/subctl/diagnose.go]: Need one more approval on #234 - can you review?

**Comment by tpantelis** [comment]
We should also add this for [diagnose all](https://github.com/submariner-io/subctl/blob/561cb4f287b2cdf5f86adf5508f24b9fa820ee14/cmd/subctl/diagnose.go#L270).

File: cmd/subctl/diagnose.go
Comment (LiavSlama) [reply cmd/subctl/diagnose.go]: Sorry, I'm a bit new with GitHub. Is this message intended for me? If so, what is needed?

Comment (tpantelis) [reply cmd/subctl/diagnose.go]: Yea it was.  I meant to review my PR as I’ve done with yours, ie click the Files changed tab and review. You can add line level comments. Then submit the review and select Approve if it looks ok

File: pkg/diagnose/servicediscovery.go
Comment (yboaron) [inline pkg/diagnose/servicediscovery.go]: What if len(se.Status.Conditions) is 0 ?

**Comment by yboaron** [comment]
> We should also add this for [diagnose all](https://github.com/submariner-io/subctl/blob/561cb4f287b2cdf5f86adf5508f24b9fa820ee14/cmd/subctl/diagnose.go#L270).

cc  @LiavSlama

File: cmd/subctl/diagnose.go
Comment (tpantelis) [inline cmd/subctl/diagnose.go]: We should check if service discovery is installed. Also I think we should move this after all the connectivity components to L291 below.

```suggestion
        if clusterInfo.ServiceDiscovery != nil {
            success = diagnose.ServiceDiscovery(clusterInfo, status) && success
         } else {
	    status.Success("Service discovery is not installed - skipping")
	}
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]:
```suggestion
			status.Failure("Error converting ServiceExport: %v", err)
			continue
```

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go:60]:
```suggestion
		status.Failure("Error listing ServiceExport resources: %v", err)
		return
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]: We shouldn't assume the `ServiceExportValid` condition is the last one as we could add other types in the future. Instead we should find the desired condition. Also remove lines 94-100 below.

**Note**: need to import "k8s.io/apimachinery/pkg/api/meta"

```suggestion
		serviceExportValid := meta.FindStatusCondition(se.Status.Conditions, mcsv1a1.ServiceExportValid)
                 if serviceExportValid == nil {
                      status.Failure("ServiceExport for %s/%s is missing the %s condition type", ns, name mcsv1a1.ServiceExportValid)
                      continue
                 }
                 
                 if serviceExportValid.Status != corev1.ConditionTrue {
                     status.Failure("Export failed or has not yet completed for service %s/%s: %s", ns, name, serviceExportValid.Message)
                     continue
                 }
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/diagnose/servicediscovery.go
Comment (LiavSlama) [reply pkg/diagnose/servicediscovery.go]: Thanks @tpantelis ! I get an IncompatibleAssign in L76 for serviceExportValid in both fields when trying to use this. Do you have any idea what I can use instead?

Comment (tpantelis) [reply pkg/diagnose/servicediscovery.go]: `ServiceExport` defines its own Condition type instead of using the standard one in `metav1` - that's unfortunate.  So can't use `meta.FindStatusCondition` - you'll have to copy the code and adjust it - it's not much anyway.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/diagnose/servicediscovery.go
Comment (tpantelis) [inline pkg/diagnose/servicediscovery.go]: It's simpler if `serviceExportValid` is a pointer.

```suggestion
		var serviceExportValid *mcsv1a1.ServiceExportCondition
		for i := range conditions {
			if conditions[i].Type == mcsv1a1.ServiceExportValid {
				serviceExportValid = &conditions[i]
				break
			}
		}

		if serviceExportValid == nil {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Review by yboaron — APPROVED** [review]

**Comment by aswinsuryan** [comment]
The lighhouse coredns configuration needs to be validated as well, 

1) If the lighthouse the configuration map exists
2) The kube-dns config map is configured to forward all requests to lighthouse cordns for the domain "clusterset.local"

This can be done in a separate PR, but the issue shall be kept open for it.

**Review by aswinsuryan — APPROVED** [review]

**Comment by vthapar** [comment]
> The lighhouse coredns configuration needs to be validated as well,
> 
> 1. If the lighthouse the configuration map exists
> 2. The kube-dns config map is configured to forward all requests to lighthouse cordns for the domain "clusterset.local"
> 
> This can be done in a separate PR, but the issue shall be kept open for it.

That should also be done as part of deployment check.

**Review by vthapar — CHANGES_REQUESTED** [review]
Changes look good, but we need to modify subctl e2e to actually test this code. @tpantelis We don't export any services as part of subctl diagnose test, right? Which means the export test will likely be skipped as nothing is exported. So we need to create a service and export it before running diagnose in e2e.

**Comment by tpantelis** [comment]
> Changes look good, but we need to modify subctl e2e to actually test this code. @tpantelis We don't export any services as part of subctl diagnose test, right? Which means the export test will likely be skipped as nothing is exported. So we need to create a service and export it before running diagnose in e2e.

Yeah I was thinking that too. We should export a service not only for diagnose but for other commands as well like gather etc. But I was thinking we could do that in another PR. I can look into it.

**Comment by tpantelis** [comment]
> > Changes look good, but we need to modify subctl e2e to actually test this code. @tpantelis We don't export any services as part of subctl diagnose test, right? Which means the export test will likely be skipped as nothing is exported. So we need to create a service and export it before running diagnose in e2e.
> 
> Yeah I was thinking that too. We should export a service not only for diagnose but for other commands as well like gather etc. But I was thinking we could do that in another PR. I can look into it.

Pushed https://github.com/submariner-io/subctl/pull/261

**Review by vthapar — APPROVED** [review]

**Comment by yboaron** [comment]
I think we can merge this PR, subctl/system_tests fails on a known issue (submariner-gateway-metrics test)  unrelated to this PR.

**Pull request merged**
